pulp_container/app/models.py

```diff
@@ -188,7 +195,63 @@ class Meta:
         unique_together = (("name", "tagged_manifest"),)
 
 
-class ContainerNamespace(BaseModel, AutoAddObjPermsMixin, AutoDeleteObjPermsMixin):
+class CreateGroupAndAddPermissionsMixin(AutoAddObjPermsMixin):
+    """
+    A mixin that creates a Group with specified permissions.
+
+    This mixin provides the ``create_group`` which creates groups with object level permissions.
+    """
+
+    def create_group(self, permissions, parameters):
+        """
+        Creates a group and optionally adds the current user to it.
+
+        The parameters are specified as a dictionary with the following keys:
+
+        "name_prefix" - the first part of the group name
+        "field_name" - the name of the model field that should be used to obtain the second part
+                       of the group name.
+        "add_user_to_group" - a boolean that specifies if the current user should be added to the
+                              group.the "model_field" for the instance.
+
+        The permissions are object level permissions assigned to the group.
+        """
+
+        prefix = parameters["name_prefix"]
+        group_type = parameters["group_type"]
+        field_name = parameters["field_name"]
+        add_user_to_group = parameters["add_user_to_group"]
+        group = Group.objects.create(
+            name="{}.{}.{}".format(prefix, group_type, getattr(self, field_name))
+        )
+        current_user = get_current_authenticated_user()
+        owners_group = Group.objects.get(
+            name="{}.{}.{}".format(prefix, "owners", getattr(self, field_name))
+        )
+        assign_perm("auth.change_group", owners_group, group)
+        assign_perm("auth.view_group", owners_group, group)
+        if add_user_to_group:
+            current_user.groups.add(group)
+        self.add_for_groups(permissions, group.name)
+
+
+class AutoDeleteGroupsMixin:
+    """
+    A mixin that automatically deletes Groups for an object prior to deletion.
+
+    To use this mixin, your model must support ``django-lifecycle``.
+    """
+
+    @hook("before_delete")
+    def delete_groups_and_user_obj_perms(self):
+        """Delete all groups associated with this object and user object permissions."""
+
+        group_pks = [o.group_id for o in GroupObjectPermission.objects.filter(object_pk=self.pk)]
+        Group.objects.filter(pk__in=group_pks).delete()
```

Comment:
This will delete all groups with any permission on that object. Even the ones not created by us.

Reply:
I'll add better filtering for this

Reply:
How about deleting only groups which are created automatically by us? all the default groups: owner, collaborator, consumer?
I'm not sure we want to touch groups created by the users, even if the only permission which is left is for the object which just got removed.
Imagine I have a group of users, teamA, and I give them only read access to a specific object. When this object is removed, they no longer have access to anything but that's fine, I don't want to recreate the group and add all the users again, I just want to give them new permissions eventually.

Reply:
I agree, let's auto remove only groups we have created. Rest of the groups would need to be deleted by the owner of the groups 'manually'